engine/access/rest/routes/websocket_handler.go

```diff
@@ -0,0 +1,283 @@
+// It listens to the subscription's channel for events and writes them to the WebSocket connection.
+// If an error occurs or the subscription channel is closed, it handles the error or termination accordingly.
+// The function uses a ticker to periodically send ping messages to the client to maintain the connection.
+func (wsController *WebsocketController) writeEvents(sub state_stream.Subscription) {
+	ticker := time.NewTicker(pingPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case _, ok := <-wsController.readChannel:
+			// we use `readChannel` as indicator of client's status, when `readChannel` closes it means that client
+			// connection has been terminated and we need to stop this goroutine to avoid memory leak.
+			if !ok {
+				return
+			}
+		case event, ok := <-sub.Channel():
+			if !ok {
+				if sub.Err() != nil {
+					err := fmt.Errorf("stream encountered an error: %v", sub.Err())
+					wsController.wsErrorHandler(err)
+					return
+				}
+				err := fmt.Errorf("subscription channel closed, no error occurred")
+				wsController.wsErrorHandler(models.NewRestError(http.StatusRequestTimeout, "subscription channel closed", err))
+				return
+			}
+			err := wsController.conn.SetWriteDeadline(time.Now().Add(writeWait))
+			if err != nil {
+				wsController.wsErrorHandler(models.NewRestError(http.StatusInternalServerError, "failed to set the initial write deadline error: ", err))
+				return
+			}
+			// Write the response to the WebSocket connection
+			err = wsController.conn.WriteJSON(event)
+			if err != nil {
+				wsController.wsErrorHandler(err)
+				return
+			}
+		case <-ticker.C:
+			err := wsController.conn.SetWriteDeadline(time.Now().Add(writeWait))
+			if err != nil {
+				wsController.wsErrorHandler(models.NewRestError(http.StatusInternalServerError, "failed to set the initial write deadline error: ", err))
+				return
+			}
+			if err := wsController.conn.WriteMessage(websocket.PingMessage, nil); err != nil {
+				wsController.wsErrorHandler(err)
+				return
+			}
+		}
+	}
+}
+
+// read function handles WebSocket messages from the client.
+// It continuously reads messages from the WebSocket connection and closes
+// the associated read channel when the connection is closed by client or when an
+// any additional message is received from the client.
+//
+// This method should be called after establishing the WebSocket connection
+// to handle incoming messages asynchronously.
+func (wsController *WebsocketController) read() {
+	// Start a goroutine to handle the WebSocket connection
+	defer close(wsController.readChannel) // notify websocket about closed connection
+
+	for {
+		// reads messages from the WebSocket connection when the connection is closed by client or when an
+		// 1) when the connection is closed by client
+		// 2) when an any additional message is received from the client
+		_, msg, err := wsController.conn.ReadMessage()
+		if err != nil {
+			return
```

Comment:
after testing this out, I think it would be helpful to update `wsController.readChannel` to be receive errors, and then have each of these cases return the error. 

then in the `writeEvents` loop, return the error is one is received. This would make it easier for clients to debug.